setup.py

```diff
@@ -1,8 +1,163 @@
 import site
 import sys
+import os
 from setuptools import setup
+from setuptools.command.build_py import build_py
+from pkgconfig import pkgconfig
+from pycparser import c_parser, preprocess_file
+from pycparser.c_ast import Decl, Typedef, TypeDecl, IdentifierType, Struct, ArrayDecl
 
 # workaround bug https://github.com/pypa/pip/issues/7953
 site.ENABLE_USER_SITE = "--user" in sys.argv[1:]
 
-setup(use_scm_version=True, cffi_modules=["scripts/libtss2_build.py:ffibuilder"])
+
+class type_generator(build_py):
+    cares = set(
+        (
+            "TPM2_ALG_ID",
+            "TPM2_ST",
+            "TPM2_ECC_CURVE",
+            "TPM2_CC",
+            "TPM2_CAP",
+            "TPM2_PT",
+            "TPM2_PT_PCR",
+            "TPMA_SESSION",
+            "TPMA_LOCALITY",
+            "TPMA_NV",
+            "TPMA_CC",
+            "TPMA_OBJECT",
+            "TPMA_ALGORITHM",
+            "TPM2_HANDLE",
+        )
+    )
+
+    type_mapping = {
+        "TPM2_ALG_ID": "TPM2_ALG",
+        "TPMI_RH_HIERARCHY": "TPM2_RH",
+        "TPMI_RH_ENABLES": "TPM2_RH",
+        "TPMI_RH_HIERARCHY_AUTH": "TPM2_RH",
+        "TPMI_RH_HIERARCHY_POLICY": "TPM2_RH",
+        "TPMI_RH_PLATFORM": "TPM2_RH",
+        "TPMI_RH_OWNER": "TPM2_RH",
+        "TPMI_RH_ENDORSEMENT": "TPM2_RH",
+        "TPMI_RH_PROVISION": "TPM2_RH",
+        "TPMI_RH_CLEAR": "TPM2_RH",
+        "TPMI_RH_NV_AUTH": "TPM2_RH",
+        "TPMI_RH_LOCKOUT": "TPM2_RH",
+        "TPMI_RH_NV_INDEX": "TPM2_RH",
+        "TPMI_RH_AC": "TPM2_RH",
+        "TPMI_RH_ACT": "TPM2_RH",
+    }
```

Comment:
Why the cares and the type_mapping dictionaries? What information do we need that we're missing?

Reply:
cares are the typedefs that it checks for (including child typedefs), type_mapping is overriding some, for example TPMI_RH_OWNER is a typedef of TPM2_HANDLE, but I found TPM2_RH to be a better match

Reply:
We want it to match what the original typedef is from the C source, no?
`tss2_tpm2_types.h:850:typedef TPM2_HANDLE TPMI_RH_OWNER;`

Also, if we're looking at a struct to get it's original type, isn't that "cares". I don't still get why we need to maintain a static map. In my head the logic is:
1. look at struct/union
2. look at sub fields
3. for each subfield, if it's a primitive type, add the mapping to the original typedef. If it's a struct/union, recurse to step 1.

Reply:
As most TPMI_RH_ types reflect a TPM2_RH constants, or a list of possible TPM2_RH constants.
By mapping it to TPM2_RH calls like str(ticket.hierarchy) or f"ticket hierachy: {ticket.hierarchy}" will work in an expected way.